src/crimson/osd/object_context.h

```diff
@@ -128,10 +128,21 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
   template <typename Lock, typename Func>
   auto _with_lock(Lock& lock, Func&& func) {
     Ref obc = this;
-    return lock.lock().then([&lock, func = std::forward<Func>(func), obc]() mutable {
-      return seastar::futurize_invoke(func).finally([&lock, obc] {
-	lock.unlock();
-      });
+    auto maybe_fut = lock.lock();
+    return seastar::futurize_invoke([
+        maybe_fut=std::move(maybe_fut),
+        func=std::forward<Func>(func)]() mutable {
+      if (maybe_fut) {
+        return std::move(*maybe_fut
+        ).then([func=std::forward<Func>(func)]() mutable {
+          return seastar::futurize_invoke(func);
+        });
+      } else {
+        // atomically calling func upon locking
+        return seastar::futurize_invoke(func);
```

Comment:
General question, is it safe to assume that avoiding chaining `lock()` returned will not cause seastar's scheduler to let other continuation run (i.e atomically calling func)? Is the same true for multi core?

Reply:
@cyx1231st, Note: While testing my PR I noticed: ```Assertion `readers == 1' failed```.
The solution here should resolve this case as well as `try_lock_for_read()` is also called atomically.

Reply:
> In the scenario where both the first and the second writers were added to waiters, once they pop one after each other - the first writer still won't promote from write atomically.
Am I missing something?

The assumption here is the first user of the obc can promote and load the obc atomically, and only the first user need to load the obc (with promotion).

In another word, if promote from write happens, the caller must be the first user and the writer (writers == 1), and if promote from read happens, the the caller must be the first user and the reader (readers == 1).

Reply:
> The assumption here is the first user of the obc can promote and load the obc atomically, and only the first user need to load the obc (with promotion).

Ah, so even if there are ongoing readers and the writers will need to wait, we can assume that the waiters won't need to load the obc (with promotion) as this is always the **first** user (reader/writer) responsibility.
Thanks for clarifying.

Reply:
> General question, is it safe to assume that avoiding chaining lock() returned will not cause seastar's scheduler to let other continuation run (i.e atomically calling func)?

Without chaining the future (or with coroutine, co_await/co_yield), it is impossible to schedule another continuation in the middle, because there is only one thread (per core) and the resources are not shared between the cores. In a specific core, the implementation is single-threaded and task scheduling is cooperative (i.e. user knows/controls where the task scheduling can happen)

> Is the same true for multi core?

True, see above.